### ISSUE TITLE
Get different  timesteps from sequence tensor

### ISSUE THREAD

**Issue by maxbry** [fchollet/keras#2567]
It is common to use a structure like this when working with encoder-decoder networks:

```
output = GRU(hidden_size, dropout_W=dropout, dropout_U=dropout, return_sequences=False)(input)
output = RepeatVector(10)(output)
```

What this does is returning the last state of the GRU (`X[:, -1, :]`). What I want is to not get the last state for every batch entry but to get the state at different timesteps for each batch entry. I tried this:

```
indexes = Input((None,) dtype='int32')
output = GRU(hidden_size, dropout_W=dropout, dropout_U=dropout, return_sequences=True)(input)
output = output[:, indexes, :]
output = RepeatVector(10)(output)
```

This is what I get:
`Exception: Input 0 is incompatible with layer repeatvector_1: expected ndim=2, found ndim=3`

What do I have to change to create a two-dimensional object?

**Comment by braingineer** [comment]
Your code is really jumbled.  Can you clarify what you mean?  Do you want to get the state at **very specific** time steps? Aka, if your data is `(batch, time, feature)` then you are saying you want to get a couple specific slices from the `time` dimension?  

What is happening with the repeat vector?  That isn't clear.  Repeat will take a 2 dimensional matrix, expand the dimensions so a new dimension is inserted at the `axis=1` and repeat it N times across that axis. (which you can see as broadcasting the matrix across the time dimension).  

Getting slices off a single dimension is easy.  Getting a set of indices takes a bit more work.  I have some code that does this, though, if this is what  you mean.

**Comment by maxbry** [comment]
I updated my comment and made the code more readable. And yes, your assumption is correct: I want to get a specific slice in the time-dimension for every batch.

What is happening with the repeat vector is the same as usual. Normally, I would set `return_sequences=False` on my last LSTM/GRU to get a two-dimensional object with the features of the last timestept and with shape `(batches, feature)`. Now, I don't want to have the features of the last timestep but for given timesteps.

**Comment by braingineer** [comment]
Ah k, that makes sense.  What do you want your final shape to be?  Aka, which dim do you want the select indices to be on.  

`(batch, selected_indices, repeated_dim, features)`
or
`(batch, repeated_dim, selected_indices, features`

To accomplish this and package it, you could write a function for a Lambda layer, or an actual layer.  Performing matrix calculations with keras tensors outside of layers is always a bad idea. the Layer class takes special care to annotate tensors with shape and history information.  So if you do matrix calculations outside of a layer, you lose that information.

**Comment by maxbry** [comment]
The shape after slicing with my indexes should be the same as if I use `return_sequences=False` on a recurrent layer: `(batch, features)`.

The recurrent layer returns `(batch, timestep, feature)` when I set `return_sequences=True`, after which I slice with the given indexes to get the same shape as with `return_sequences=False`.

**Comment by braingineer** [comment]
> after which I slice with the given indexes to get the same shape as with return_sequences=False.

Not quite. Starting from `(batch, time, feature)`,  when you pass in indices, you will get a middle dimension the shape of the number of indices you passed in.  If you index into the middle dimension with 3 indices.. where do the 3 slices go? Are you assuming they get combined in some way?  Why would it just suddenly become ```(batch, feature)```

```python
A = np.arange(12).reshape(2,3,2)
print(A)
'''
[[[ 0  1]
  [ 2  3]
  [ 4  5]]

 [[ 6  7]
  [ 8  9]
  [10 11]]]
'''
A2 = A[:,np.array([0,1]),:]
print(A2)
'''
[[[0 1]
  [2 3]]

 [[6 7]
  [8 9]]]
'''
print(A2.shape)
'''
(2, 2, 2)
'''
```

**Comment by maxbry** [comment]
Well, this may be the reason why my approach isn't working. But what does work is:

    >>> A = np.zeros((2,3,2))
    >>> A[[0,1], [0,1], :].shape
    (2, 2)

**Comment by joelthchao** [comment]
@maxbry I can only provide a possible way for TensorFlow
```python
input = Input(batch_shape=(batch_size, timestep, features))
index = Input(batch_shape=(1,), dtype='int32')
index = K.concatenate([[0], index, [0]], axis=0)
output = GRU(hidden_size, dropout_W=dropout, dropout_U=dropout, return_sequences=True)(input)
output = tf.slice(output, index, [batch_size, 1, features])
```

**Comment by braingineer** [comment]
ahh. so, you want specific indices from specific rows.  

I have two examples below.  One is a column gather for Theano.  The other is a layer I use for dynamic embeddings on the time dimension. 

```python
def colgather(reference, col_indices, row_indices=None):
    '''select indices from the specified cols in 1st dim for specified rows in 0th dim

    # Arguments
    reference: a tensor 
    col_indices: a set of int indices; 
                 If row_indices is not specified, it should have the shape property
    row_indices: (optional) a set of int indices;
                 Has to match the size of col_indices

    # Returns
        Tensor with columns selected and 1 less ndim
    '''
    row_indices = row_indices or T.arange(col_indices.shape[0])
    indices = tuple([row_indices, col_indices] + [slice(None) for _ in range(2,reference.ndim)])
    return reference[indices]
```



```python
class DynamicEmbedding(Embedding):
    def __init__(self, embedding_matrix, mode='matrix', *args, **kwargs):
        assert hasattr(embedding_matrix, '_keras_shape')
        self.W = embedding_matrix
        if mode=='tensor':
            assert len(embedding_matrix._keras_shape) == 3
            indim = self.W._keras_shape[1]
            outdim = self.W._keras_shape[2]
        else:
            assert len(embedding_matrix._keras_shape) == 2
            indim, outdim = self.W._keras_shape

        self.mode = mode
        super(DynamicEmbedding, self).__init__(indim, outdim, *args, **kwargs)
        
    def build(self, input_shape):
        self.constraints = {}
        if self.W_constraint:
            self.constraints[self.W] = self.W_constraint

        self.regularizers = []
        if self.W_regularizer:
            self.W_regularizer.set_param(self.W)
            self.regularizers.append(self.W_regularizer)

        if self.activity_regularizer:
            self.activity_regularizer.set_layer(self)
            self.regularizers.append(self.activity_regularizer)

        if self.initial_weights is not None:
            self.set_weights(self.initial_weights)
            
    def call(self, x, mask=None):
        if 0. < self.dropout < 1.:
            retain_p = 1. - self.dropout
            dims = self.W._keras_shape[:-1]
            B = K.random_binomial(dims, p=retain_p) * (1. / retain_p)
            B = K.expand_dims(B)
            W = K.in_train_phase(self.W * B, self.W)
        else:
            W = self.W
        
        if self.mode == 'matrix':
            return K.gather(W,x)
        elif self.mode == 'tensor':
            # quick and dirty: only allowing for 3dim inputs when it's tensor mode
            assert K.ndim(x) == 3
            # put sequence on first; gather; take diagonal across shared batch dimension
            # in other words, W is (B, S, F)
            # incoming x is (B, S, A)
            inds = K.arange(self.W._keras_shape[0])
            #out = K.gather(K.permute_dimensions(W, (1,0,2)), x).diagonal(axis1=0, axis2=3)
            #return K.permute_dimensions(out, (3,0,1,2))
            ### method above doesn't do grads =.=
            out = K.gather(K.permute_dimensions(W, (1,0,2)), x)
            out = K.permute_dimensions(out, (0,3,1,2,4))
            out = K.gather(out, (inds, inds))
            return out
        else:
            raise Exception('sanity check. should not be here.')

**Comment by Youmna-H** [comment]
I am trying to do something similar now and get the output at a certain timestep and feed it to another layer.
lstm = LSTM(100, return_sequences=True)
and then do something like:
Convolution1D(nb_filter=100, filter_length=3)(lstm[-1])
to only use the final timestep.
I don't want to use 'return_sequences=False' as I need the other timesteps too.
Was that resolved?
